yggdrasilffi/src/lib.rs

```diff
@@ -194,6 +195,23 @@ pub unsafe extern "C" fn take_state(
     result_to_json_ptr(result)
 }
 
+/// Gets the current state of the engine as a JSON encoded `ClientFeatures` structure.
+///
+/// # Safety
+///
+/// This function dereferences the engine_ptr argument, and so this function should not
+/// be called with a null pointer.
+#[no_mangle]
+pub unsafe extern "C" fn get_state(engine_ptr: *mut c_void) -> *const c_char {
+    let result: Result<Option<ClientFeatures>, FFIError> = (|| {
+        let guard = get_engine(engine_ptr)?;
+        let engine = recover_lock(&guard);
+        Ok(Some(engine.get_state()))
```

Comment:
Haven't tried it but I want to say if we write this to a JSON string on this line we can return a string here and then the host binding languages don't have to write that back to a string.

More a matter of taste than anything else. This isn't a hot path so the perf doesn't matter here and this approach will absolutely work. It would be nice if the WASM and FFI layer did the same thing though